// @tarojs/mini-runner/dist/template/adapters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const weixinAdapter = {
    if: 'wx:if',
    else: 'wx:else',
    elseif: 'wx:elif',
    for: 'wx:for',
    forItem: 'wx:for-item',
    forIndex: 'wx:for-index',
    key: 'wx:key',
    xs: 'wxs',
    type: "weapp" /* WEAPP */
};
const swanAdapter = {
    if: 's-if',
    else: 's-else',
    elseif: 's-elif',
    for: 's-for',
    forItem: 's-for-item',
    forIndex: 's-for-index',
    key: 's-key',
    xs: 'sjs',
    type: "swan" /* SWAN */
};
const alipayAdapter = {
    if: 'a:if',
    else: 'a:else',
    elseif: 'a:elif',
    for: 'a:for',
    forItem: 'a:for-item',
    forIndex: 'a:for-index',
    key: 'a:key',
    xs: 'sjs',
    type: "alipay" /* ALIPAY */
};
const ttAdapter = {
    if: 'tt:if',
    else: 'tt:else',
    elseif: 'tt:elif',
    for: 'tt:for',
    forItem: 'tt:for-item',
    forIndex: 'tt:for-index',
    key: 'tt:key',
    type: "tt" /* TT */
};
const quickappAdapter = {
    if: 'if',
    else: 'else',
    elseif: 'elif',
    for: 'for',
    forItem: 'for-item',
    forIndex: 'for-index',
    key: 'key',
    type: "quickapp" /* QUICKAPP */
};
const qqAdapter = {
    if: 'qq:if',
    else: 'qq:else',
    elseif: 'qq:elif',
    for: 'qq:for',
    forItem: 'qq:for-item',
    forIndex: 'qq:for-index',
    key: 'qq:key',
    xs: 'wxs',
    type: "qq" /* QQ */
};
const jdAdapter = {
    if: 'jd:if',
    else: 'jd:else',
    elseif: 'jd:elif',
    for: 'jd:for',
    forItem: 'jd:for-item',
    forIndex: 'jd:for-index',
    key: 'jd:key',
    type: "jd" /* JD */
};
exports.supportXS = () => {
    return ["qq" /* QQ */, "weapp" /* WEAPP */, "swan" /* SWAN */, "alipay" /* ALIPAY */].includes(exports.Adapter.type);
};
exports.Adapter = weixinAdapter;
function setAdapter(adapter) {
    switch (adapter.toLowerCase()) {
        case "swan" /* SWAN */:
            exports.Adapter = swanAdapter;
            break;
        case "alipay" /* ALIPAY */:
            exports.Adapter = alipayAdapter;
            break;
        case "tt" /* TT */:
            exports.Adapter = ttAdapter;
            break;
        case "quickapp" /* QUICKAPP */:
            exports.Adapter = quickappAdapter;
            break;
        case "qq" /* QQ */:
            exports.Adapter = qqAdapter;
            break;
        case "jd" /* JD */:
            exports.Adapter = jdAdapter;
            break;
        default:
            exports.Adapter = weixinAdapter;
            break;
    }
}
exports.setAdapter = setAdapter;
