// @tarojs/mini-runner/dist/webpack/build.conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const runner_utils_1 = require("@tarojs/runner-utils");
const chain_1 = require("./chain");
const base_conf_1 = require("./base.conf");
const MiniPlugin_1 = require("../plugins/MiniPlugin");
const emptyObj = {};
exports.default = (appPath, mode, config) => {
    const chain = base_conf_1.default(appPath);
    const { buildAdapter = "weapp" /* WEAPP */, alias = emptyObj, entry = emptyObj, output = emptyObj, outputRoot = 'dist', sourceRoot = 'src', designWidth = 750, deviceRatio, enableSourceMap = process.env.NODE_ENV !== 'production', baseLevel = 16, framework = 'nerv', prerender, defineConstants = emptyObj, env = emptyObj, cssLoaderOption = emptyObj, sassLoaderOption = emptyObj, lessLoaderOption = emptyObj, stylusLoaderOption = emptyObj, mediaUrlLoaderOption = emptyObj, fontUrlLoaderOption = emptyObj, imageUrlLoaderOption = emptyObj, miniCssExtractPluginOption = emptyObj, postcss = emptyObj, nodeModulesPath, quickappJSON, csso, terser, commonChunks, addChunkPages } = config;
    let { copy } = config;
    const plugin = {};
    const minimizer = [];
    const sourceDir = path.join(appPath, sourceRoot);
    const outputDir = path.join(appPath, outputRoot);
    const taroBaseReg = /@tarojs[\/][a-z]+/;
    if (config.isBuildPlugin) {
        const patterns = copy ? copy.patterns : [];
        patterns.push({
            from: path.join(sourceRoot, 'plugin', 'doc'),
            to: path.join(outputRoot, 'doc')
        });
        patterns.push({
            from: path.join(sourceRoot, 'plugin', 'plugin.json'),
            to: path.join(outputRoot, 'plugin', 'plugin.json')
        });
        copy = Object.assign({}, copy, { patterns });
    }
    if (copy) {
        plugin.copyWebpackPlugin = chain_1.getCopyWebpackPlugin({ copy, appPath });
    }
    if (framework === runner_utils_1.FRAMEWORK_MAP.VUE) {
        const VueLoaderPlugin = require('vue-loader/lib/plugin');
        plugin.vueLoaderPlugin = {
            plugin: new VueLoaderPlugin()
        };
    }
    alias[runner_utils_1.taroJsComponents] = `${runner_utils_1.taroJsComponents}/mini`;
    if (framework === 'react') {
        alias['react-dom'] = '@tarojs/react';
    }
    if (framework === 'nerv') {
        alias['react-dom'] = 'nervjs';
        alias.react = 'nervjs';
    }
    env.FRAMEWORK = JSON.stringify(framework);
    env.TARO_ENV = JSON.stringify(buildAdapter);
    const constantsReplaceList = chain_1.mergeOption([chain_1.processEnvOption(env), defineConstants]);
    const entryRes = chain_1.getEntry({
        sourceDir,
        entry,
        isBuildPlugin: config.isBuildPlugin
    });
    const defaultCommonChunks = !!config.isBuildPlugin
        ? ['plugin/runtime', 'plugin/vendors', 'plugin/taro', 'plugin/common']
        : ['runtime', 'vendors', 'taro', 'common'];
    let customCommonChunks = defaultCommonChunks;
    if (typeof commonChunks === 'function') {
        customCommonChunks = commonChunks(defaultCommonChunks.concat()) || defaultCommonChunks;
    }
    else if (Array.isArray(commonChunks) && commonChunks.length) {
        customCommonChunks = commonChunks;
    }
    plugin.definePlugin = chain_1.getDefinePlugin([constantsReplaceList]);
    plugin.miniPlugin = chain_1.getMiniPlugin({
        sourceDir,
        outputDir,
        buildAdapter,
        constantsReplaceList,
        nodeModulesPath,
        quickappJSON,
        designWidth,
        pluginConfig: entryRes.pluginConfig,
        isBuildPlugin: !!config.isBuildPlugin,
        commonChunks: customCommonChunks,
        baseLevel,
        framework,
        prerender,
        addChunkPages
    });
    plugin.miniCssExtractPlugin = chain_1.getMiniCssExtractPlugin([{
            filename: `[name]${runner_utils_1.MINI_APP_FILES[buildAdapter].STYLE}`,
            chunkFilename: `[name]${runner_utils_1.MINI_APP_FILES[buildAdapter].STYLE}`
        }, miniCssExtractPluginOption]);
    plugin.providerPlugin = chain_1.getProviderPlugin({
        window: ['@tarojs/runtime', 'window'],
        document: ['@tarojs/runtime', 'document'],
        navigator: ['@tarojs/runtime', 'navigator']
    });
    const isCssoEnabled = !((csso && csso.enable === false));
    const isTerserEnabled = !((terser && terser.enable === false));
    if (mode === 'production') {
        if (isTerserEnabled) {
            minimizer.push(chain_1.getTerserPlugin([
                enableSourceMap,
                terser ? terser.config : {}
            ]));
        }
        if (isCssoEnabled) {
            const cssoConfig = csso ? csso.config : {};
            plugin.cssoWebpackPlugin = chain_1.getCssoWebpackPlugin([cssoConfig]);
        }
    }
    chain.merge({
        mode,
        devtool: chain_1.getDevtool(enableSourceMap),
        entry: entryRes.entry,
        output: chain_1.getOutput(appPath, [{
                outputRoot,
                publicPath: '/',
                buildAdapter
            }, output]),
        target: MiniPlugin_1.Targets[buildAdapter],
        resolve: { alias },
        module: chain_1.getModule(appPath, {
            sourceDir,
            buildAdapter,
            // constantsReplaceList,
            designWidth,
            deviceRatio,
            enableSourceMap,
            cssLoaderOption,
            lessLoaderOption,
            sassLoaderOption,
            stylusLoaderOption,
            fontUrlLoaderOption,
            imageUrlLoaderOption,
            mediaUrlLoaderOption,
            postcss
            // babel
        }),
        plugin,
        optimization: {
            usedExports: true,
            minimizer,
            runtimeChunk: {
                name: config.isBuildPlugin ? 'plugin/runtime' : 'runtime'
            },
            splitChunks: {
                chunks: 'all',
                maxInitialRequests: Infinity,
                minSize: 0,
                cacheGroups: {
                    common: {
                        name: !!config.isBuildPlugin ? 'plugin/common' : 'common',
                        minChunks: 2,
                        priority: 1
                    },
                    vendors: {
                        name: !!config.isBuildPlugin ? 'plugin/vendors' : 'vendors',
                        minChunks: 2,
                        test: module => {
                            return /[\\/]node_modules[\\/]/.test(module.resource);
                        },
                        priority: 10
                    },
                    taro: {
                        name: !!config.isBuildPlugin ? 'plugin/taro' : 'taro',
                        test: module => {
                            return taroBaseReg.test(module.context);
                        },
                        priority: 100
                    }
                }
            }
        }
    });
    return chain;
};
