// @tarojs/components/dist/esm-es5/taro-radio-group-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement, H as Host } from './core-42efd82a.js';
var RadioGroup = /** @class */ (function () {
    function RadioGroup(hostRef) {
        registerInstance(this, hostRef);
        this.uniqueName = Date.now().toString(36);
        this.onChange = createEvent(this, "change", 7);
    }
    RadioGroup.prototype.function = function (e) {
        e.stopPropagation();
        if (e.target.tagName !== 'TARO-RADIO')
            return;
        var target = e.target;
        if (target.checked) {
            var childList = this.el.querySelectorAll('taro-radio-core');
            childList.forEach(function (element) {
                if (element !== target) {
                    element.checked = false;
                }
            });
            this.value = e.detail.value;
            this.onChange.emit({
                value: this.value
            });
        }
    };
    RadioGroup.prototype.componentDidLoad = function () {
        var _this = this;
        var childList = this.el.querySelectorAll('taro-radio-core');
        childList.forEach(function (element) {
            element.setAttribute('name', _this.name || _this.uniqueName);
        });
        Object.defineProperty(this.el, 'value', {
            get: function () { return _this.value; },
            configurable: true
        });
    };
    RadioGroup.prototype.render = function () {
        return (h(Host, { class: 'weui-cells_radiogroup' }));
    };
    Object.defineProperty(RadioGroup.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    return RadioGroup;
}());
export { RadioGroup as taro_radio_group_core };
