// @tarojs/components/dist/esm-es5/taro-checkbox-group-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement, H as Host } from './core-42efd82a.js';
var CheckboxGroup = /** @class */ (function () {
    function CheckboxGroup(hostRef) {
        registerInstance(this, hostRef);
        this.uniqueName = Date.now().toString(36);
        this.value = [];
        this.onChange = createEvent(this, "change", 7);
    }
    CheckboxGroup.prototype.function = function (e) {
        var _this = this;
        e.stopPropagation();
        if (e.target.tagName !== 'TARO-CHECKBOX')
            return;
        var childList = this.el.querySelectorAll('taro-checkbox-core');
        this.value = Array.from(childList)
            .filter(function (element) {
            var _a;
            var checkbox = element.querySelector("input[name=" + (_this.name || _this.uniqueName) + "]");
            return (_a = checkbox) === null || _a === void 0 ? void 0 : _a.checked;
        })
            .map(function (element) { return element.value; });
        this.onChange.emit({
            value: this.value
        });
    };
    CheckboxGroup.prototype.componentDidLoad = function () {
        var _this = this;
        var childList = this.el.querySelectorAll('taro-checkbox-core');
        childList.forEach(function (element) {
            element.setAttribute('name', _this.name || _this.uniqueName);
        });
        Object.defineProperty(this.el, 'value', {
            get: function () { return _this.value; },
            configurable: true
        });
    };
    CheckboxGroup.prototype.render = function () {
        return (h(Host, null));
    };
    Object.defineProperty(CheckboxGroup.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    return CheckboxGroup;
}());
export { CheckboxGroup as taro_checkbox_group_core };
