// @tarojs/components/dist/esm-es5/taro-radio-core.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-42efd82a.js';
var Radio = /** @class */ (function () {
    function Radio(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.value = '';
        this.checked = false;
        this.handleClick = function () {
            if (!_this.checked)
                _this.checked = true;
        };
        this.onChange = createEvent(this, "radiochange", 7);
    }
    Radio.prototype.watchChecked = function (newVal) {
        newVal && this.onChange.emit({ value: this.value });
    };
    Radio.prototype.watchId = function (newVal) {
        if (newVal)
            this.inputEl.setAttribute('id', newVal);
    };
    Radio.prototype.componentDidRender = function () {
        this.id && this.el.removeAttribute('id');
    };
    Radio.prototype.render = function () {
        var _this = this;
        var _a = this, checked = _a.checked, name = _a.name, value = _a.value;
        return (h(Host, { className: 'weui-cells_checkbox', onClick: this.handleClick }, h("input", { ref: function (dom) {
                if (!dom)
                    return;
                _this.inputEl = dom;
                if (_this.id)
                    dom.setAttribute('id', _this.id);
            }, type: 'radio', name: name, value: value, class: 'weui-check', checked: checked, onChange: function (e) { return e.stopPropagation(); } }), h("i", { class: 'weui-icon-checked' }), h("slot", null)));
    };
    Object.defineProperty(Radio.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Radio, "watchers", {
        get: function () {
            return {
                "checked": ["watchChecked"],
                "id": ["watchId"]
            };
        },
        enumerable: true,
        configurable: true
    });
    return Radio;
}());
export { Radio as taro_radio_core };
