// @tarojs/components/dist/esm-es5/taro-checkbox-core.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-42efd82a.js';
var Checkbox = /** @class */ (function () {
    function Checkbox(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.value = '';
        this.checked = false;
        this.handleChange = function (e) {
            e.stopPropagation();
            _this.onChange.emit({
                value: _this.value
            });
        };
        this.onChange = createEvent(this, "checkboxchange", 7);
    }
    Checkbox.prototype.watchId = function (newVal) {
        if (newVal)
            this.inputEl.setAttribute('id', newVal);
    };
    Checkbox.prototype.componentDidRender = function () {
        this.id && this.el.removeAttribute('id');
    };
    Checkbox.prototype.render = function () {
        var _this = this;
        var _a = this, checked = _a.checked, name = _a.name, color = _a.color, value = _a.value;
        return (h(Host, { className: 'weui-cells_checkbox' }, h("input", { ref: function (dom) {
                if (!dom)
                    return;
                _this.inputEl = dom;
                if (_this.id)
                    dom.setAttribute('id', _this.id);
            }, type: 'checkbox', value: value, name: name, class: 'taro-checkbox_checked', style: { color: color }, checked: checked, onChange: this.handleChange }), h("slot", null)));
    };
    Object.defineProperty(Checkbox.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Checkbox, "watchers", {
        get: function () {
            return {
                "id": ["watchId"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Checkbox, "style", {
        get: function () { return "\@charset \"UTF-8\";.taro-checkbox,.taro-checkbox_checked{position:relative;display:inline-block}.taro-checkbox_checked{min-height:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;font-size:23px;border:1px solid #d1d1d1;background-color:#fff;border-radius:3px;color:#1aad19;width:23px;height:23px;vertical-align:0;top:5px}.taro-checkbox_checked:checked:before{font-family:weui;font-style:normal;font-weight:400;font-variant:normal;text-transform:none;text-align:center;speak:none;display:inline-block;vertical-align:middle;text-decoration:inherit;content:\"\";color:inherit;font-size:inherit;position:absolute;top:50%;left:50%;-webkit-transform:translate(-50%,-48%) scale(.73);transform:translate(-50%,-48%) scale(.73)}"; },
        enumerable: true,
        configurable: true
    });
    return Checkbox;
}());
export { Checkbox as taro_checkbox_core };
