// @tarojs/components/dist/esm-es5/taro-button-core.entry.js
import { r as registerInstance, c as createEvent, h } from './core-42efd82a.js';
import { c as classNames } from './index-be61bb74.js';
var Button = /** @class */ (function () {
    function Button(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.hoverClass = 'button-hover';
        this.type = 'default';
        this.hoverStartTime = 20;
        this.hoverStayTime = 70;
        this.loading = false;
        this.formType = null;
        this.hover = false;
        this.touch = false;
        this.onTouchStart = function (e) {
            // 取消 form 内 button 跳转
            e.preventDefault();
            _this.touch = true;
            if (_this.hoverClass && !_this.disabled) {
                setTimeout(function () {
                    if (_this.touch) {
                        _this.hover = true;
                    }
                }, _this.hoverStartTime);
            }
        };
        this.onTouchEnd = function () {
            _this.touch = false;
            if (_this.hoverClass && !_this.disabled) {
                setTimeout(function () {
                    if (!_this.touch) {
                        _this.hover = false;
                    }
                }, _this.hoverStayTime);
            }
            if (_this.formType === 'submit') {
                _this.onSubmit.emit();
            }
            else if (_this.formType === 'reset') {
                _this.onReset.emit();
            }
        };
        this.onSubmit = createEvent(this, "tarobuttonsubmit", 7);
        this.onReset = createEvent(this, "tarobuttonreset", 7);
    }
    Button.prototype.render = function () {
        var _a;
        var _b = this, disabled = _b.disabled, hoverClass = _b.hoverClass, type = _b.type, size = _b.size, plain = _b.plain, loading = _b.loading, hover = _b.hover;
        var cls = classNames('weui-btn', 'taro-button', (_a = {},
            _a["" + hoverClass] = hover && !disabled,
            _a["weui-btn_plain-" + type] = plain,
            _a["weui-btn_" + type] = !plain && type,
            _a['weui-btn_mini'] = size === 'mini',
            _a['weui-btn_loading'] = loading,
            _a['weui-btn_disabled'] = disabled,
            _a));
        return (h("button", { class: cls, type: type,
            // @ts-ignore: weui need plain for css selector
            plain: plain, disabled: disabled, onTouchStart: this.onTouchStart, onTouchEnd: this.onTouchEnd }, loading && h("i", { class: 'weui-loading' }), h("slot", null)));
    };
    Object.defineProperty(Button, "style", {
        get: function () { return "taro-button-core+taro-button-core button{margin-top:15px}button{position:relative;display:block;width:100%;margin-left:auto;margin-right:auto;padding-left:14px;padding-right:14px;-webkit-box-sizing:border-box;box-sizing:border-box;font-size:18px;text-align:center;text-decoration:none;line-height:2.55555556;border-radius:5px;-webkit-tap-highlight-color:transparent;overflow:hidden;color:#000;background-color:#f8f8f8}button[plain]{color:#353535;border:1px solid #353535;background-color:transparent}button[plain][disabled]{color:rgba(0,0,0,.3);border:1px solid rgba(0,0,0,.2);background-color:#f7f7f7}button[type=primary]{color:#fff;background-color:#1aad19}button[type=primary][plain]{color:#1aad19;border:1px solid #1aad19;background-color:transparent}button[type=primary][plain][disabled]{color:rgba(0,0,0,.3);border:1px solid rgba(0,0,0,.2);background-color:#f7f7f7}"; },
        enumerable: true,
        configurable: true
    });
    return Button;
}());
export { Button as taro_button_core };
