// @tarojs/components/dist/esm/taro-canvas-core.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-42efd82a.js';

const LONG_TAP_DELAY = 500;
const Canvas = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.width = 300;
        this.height = 150;
        this.onTouchStart = () => {
            this.timer = setTimeout(() => {
                this.onLongTap.emit();
            }, LONG_TAP_DELAY);
        };
        this.onTouchMove = () => {
            clearTimeout(this.timer);
        };
        this.onTouchEnd = () => {
            clearTimeout(this.timer);
        };
        this.onLongTap = createEvent(this, "longtap", 7);
    }
    componentDidLoad() {
        const { width, height } = this.el.getBoundingClientRect();
        this.width = width;
        this.height = height;
    }
    componentDidUpdate() {
        const { width, height } = this.el.getBoundingClientRect();
        if (this.width !== width)
            this.width = width;
        if (this.height !== height)
            this.height = height;
    }
    render() {
        const { canvasId, width, height } = this;
        return (h(Host, { class: 'taro-canvas' }, h("canvas", { "canvas-id": canvasId, width: width, height: height, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd })));
    }
    get el() { return getElement(this); }
    static get style() { return ".taro-canvas{position:relative;display:block;width:300px;height:150px}"; }
};

export { Canvas as taro_canvas_core };
